darwin/Platform.c

```diff
@@ -464,10 +469,58 @@ bool Platform_getDiskIO(DiskIOData* data) {
    return true;
 }
 
+/* Caution: Given that interfaces are dynamic, and it is not possible to get statistics on interfaces that no longer exist,
+   if some interface disappears between the time of two samples, the values of the second sample may be lower than those of
+   the first one. */
 bool Platform_getNetworkIO(NetworkIOData* data) {
-   // TODO
-   (void)data;
-   return false;
+   int mib[6] = {CTL_NET,
+      PF_ROUTE, /* routing messages */
+      0, /* protocal number, currently always 0 */
+      0, /* select all address families */
+      NET_RT_IFLIST2, /* interface list with addresses */
+      0};
+   size_t len;
+
+   /* Determine len */
+   if (sysctl(mib, ARRAYSIZE(mib), NULL, &len, NULL, 0) < 0)
+      return false;
+
+   char *buf = xMalloc(len);
+
+   if (sysctl(mib, ARRAYSIZE(mib), buf, &len, NULL, 0) < 0) {
+      free(buf);
+      return false;
+   }
```

Comment:
As the allocated buffer may be too small due to a race condition, try to allocate enough memory in a loop (with some slack); and only fail after 3 or so attempts to get the data.

Similar to what is done for listing processes/LWP …

Reply:
1. I wonder, is there an API that allocates the buffer and returns the `sysctl` information in one atomic step? Because it looks like we could not avoid the rece condition otherwise.
2. Is it possible that we reuse the `buf` buffer and avoid the repeated `xMalloc` and `free` routines? I mean, track the allocated buffer and it's size in global variables, and use `xRealloc` when the buffer needs expanding. I don't know if htop had used this practice before.

Reply:
> 1. I wonder, is there an API that allocates the buffer and returns the `sysctl` information in one atomic step? Because it looks like we could not avoid the race condition otherwise.

We can't fully avoid this if the interface to query the information does not provide the required size directly. Additionallythis size can change at any time, thus insufficient buffer space is not a bug in itself. Thus the code should be able to cope with the situation and increase the buffer if necessary; but break out of the loop, if repeated attempts fail. As mentioned: That's what the process information code (e.g. on Darwin) does.

> 2. Is it possible that we reuse the `buf` buffer and avoid the repeated `xMalloc` and `free` routines? I mean, track the allocated buffer and it's size in global variables, and use `xRealloc` when the buffer needs expanding. I don't know if htop had used this practice before.

That's what the `Platform_init` and `Platform_done` functions are for. Also the code should just re-allocate the buffer as needed for subsequent requests.